Add checkpoint feature

tjwatson opened the pull request:
This is a work in progress to merge the criu branch into integration.

tjwatson committed:
Add checkpoint feature

tjwatson committed:
Fix up the checkpoint feature properties

Enable feature for parallel activate

tjwatson committed:
Change snapshot to take a File

tjwatson committed:
Move checkpoint "spi" into kernel boot to allow hook impl in kernel

Without the SnapshotHook in the kernel no kernel components
can take part in the prepare/restore phase of checkpoint.
Only moving the interfaces for the "spi" to kernel, not
the implementation.  Also note that this is not real
SPI until we have a feature declare it as SPI. For
now this is a Liberty internal package

tjwatson committed:
Introduce SnapshotHookFactory

For each snapshot operation allow a factory to create new
instances of SnapshotHook so they can keep state inside their instance.

tjwatson committed:
Add unit tests for Checkpoint service

Add a ExecuteCRIU internal service interface to allow testing of calling
dump without actually doing a real criu dump.  Also let another
implementation be plugged in that does not use JNI (perhaps using gRPC)
instead.

Call abortPrepare instead of restore when an exception occurs during the
CRIU dump.

tjwatson committed:
Add criu-ns script for restoring

sebratton committed:
Changes to support launching from server script with checkpoint

tjwatson committed:
Move JNI impl of ExecuteCRIU out to separate bundle

- Make the gogo command package optional
- Add ServerReadyCheckpoint component for testing purpose

tjwatson committed:
Fix io.openliberty.criu.lib component for ExecuteCRIU_JNI

tjwatson committed:
Change the snapshot folder name to "snapshot"

tjwatson committed:
Fix hookFactories lookup name

tjwatson committed:
Add default methods and make factory a functional interface

tjwatson committed:
Add a hook to reset the start time

This is likely a temporary solution.
Look for a file in the workarea that contains
the start time in milliseconds.  This time
is then subtracted from the initial nano time
so that we can get the complete time it takes to
restore the process

tjwatson committed:
Add logging to testpath for unit tests to work

tjwatson committed:
Change phases for FEATURES and APPLICATIONS

tjwatson committed:
Rename criu.lib to jigawatts

wtlucy committed:
add entry and exit trace for H2WriteTree write queue runnable

tjwatson committed:
Get jigawatts from artifactory

sebratton committed:
code review changes

sebratton committed:
Fix launcher error message. Deprecate api for gogo

sebratton committed:
criu logging support

sebratton committed:
criu: Fix checkpoint failure when snapshot dir exists.

sebratton committed:
criu: restore and checkpoint from server shell script

Commit:
Initial drop of getAvailableProcessors changes

sebratton committed:
code review changes

sebratton committed:
Merge pull request #17560 from sebratton/criu

Criu

sebratton committed:
minor changes to criu support

add restoreTime file and suppress FFDC on error processing it

sebratton committed:
criu: add --file-locks switch to restore

tjwatson committed:
Merge pull request #17619 from sebratton/criu-update

minor changes to criu support

tjwatson committed:
Update jigawatts to return error codes from criu dump

Initially exit if criu dump returns an error.
It is unclear if we should exit only for criu dump errors
or all exceptions when doing a snapshot. For now
just simply exit when criu dump errors.

Much more work is needed to handle errors here.
For example, proper error messages and proper
FFDCs if they are appropriate.

tjwatson committed:
Merge pull request #17676 from tjwatson/criu

Update jigawatts to return error codes from criu dump

tjwatson committed:
Do system exit in a background thread

If system exit is done in the current startup thread
it will block the server from exiting while the
server is waiting for it to start

tjwatson committed:
Add org.eclipse.openj9.criu for boot delegation

This is not the ideal way to do this.  This package
is needed when running on Open J9 with CRIU support.

It would be better if the launcher would detect the
available packages from the JVM boot layer modules
similar to how the Equinox framework does when the
system packages are not set explicitly.

That is complicated to do given how we set certain
attributes on the system packages in the liberty
boot launcher.  For now just doing boot delegation
for this package.

tjwatson committed:
Use the CHECKPOINT_PROPERTY_NAME constant

tjwatson committed:
Add io.openliberty.checkpoint.openj9 bundle for checkpoint

This bundle uses proposed API in Open J9 to do a criu
dump operation.  The io.openliberty.org.eclipse.openj9.criu
project is only used to hold stubs of the Open J9 API
for compilation purposes.  This API will be provided by
future Open J9 versions.

sebratton committed:
criu updates
Move storage for criu images to workarea/checkpoint/images
Fix issue with failed restore of server launched via run command.

tjwatson committed:
Merge pull request #17740 from sebratton/criu-update

criu updates

sebratton committed:
Add java switch -XX:+EnableCRIUSupport to server. Misc cleanup.

tjwatson committed:
Merge branch 'release' into criu branch

sebratton committed:
Automaic Eclipse Java Code Style formatter changes only

Saved with latest Eclipse Java Code Style formatter in placed. No manual code changes made.

sebratton committed:
criu: add SnapshotHook for restore on .sCommand

Refactor existing code to allow reuse of code in SnapShotHook. The hook will create the .sCommand on each criu restore

sebratton committed:
Adjust file perms on .sCommand

The file perms were too permissive before.

sebratton committed:
Remove exports from bin/serer script

These were temp exports needed to operate with jigaWatts. Not needed with openJ9 support for criu

tjwatson committed:
Merge pull request #17873 from sebratton/criu-rm-server-exports

Criu: rm server exports

tjwatson committed:
Avoid major version changes to utils package

The utils package is considered SPI. This limits
what we can change to the classes in the package
TimestampUtils contains a couple of protected fields.
Even though nobody would have a valid reason to
extend TimestampUtils we still need to make sure
the protected fields do not break compatibility.

Deprecating the protected fields and adding a new
private field for the start time.

sebratton committed:
criu fixes to com.ibm.ws.kernel.boot_test

sebratton committed:
Merge pull request #17867 from sebratton/criu-update

criu: Add SnapshotHook to support multiple restores by restoring .sCommand

Commit:
Merge pull request #17552 from dettmoney/number-cpus

Calculate number of available processors at runtime

tjwatson committed:
Merge pull request #17900 from tjwatson/criuTimestampUtils

Avoid major version changes to utils package

tjwatson committed:
Fix location restoreTime is read from

tjwatson committed:
Merge pull request #17944 from tjwatson/fixRestoreTime

Fix restore time

anjumfatima90 committed:
Rework Checkpoint interface and CheckpointHook/CheckpointHookFactory classes

tjwatson committed:
Merge pull request #17964 from anjumfatima90/criu_refactor

Rework Checkpoint interface and CheckpointHook/CheckpointHookFactory …

anjumfatima90 committed:
Refactor checkpointImpl

tjwatson committed:
Fix unit tests

Unit tests do not need to check the context properties.
Intead use a constructor to set the booleans

anjumfatima90 committed:
Merge pull request #17967 from anjumfatima90/criu_refactor1

Refactor checkpointImpl

tjwatson committed:
Update unit tests to prepare to test applications/feature phases

anjumfatima90 committed:
Add unit tests for applications/features phases

anjumfatima90 committed:
Merge pull request #18045 from anjumfatima90/criu-unit-test

Criu unit test

tjwatson committed:
Initial updates to use CRIUSupport builder

sebratton committed:
criu restore.log location honors LOG_DIR env variable

sebratton committed:
Merge pull request #18134 from sebratton/update-criu

criu restore.log location honors LOG_DIR env variable

tjwatson committed:
Add symbol server.logs.dir

tjwatson committed:
Write checkpoint.log to server logs/ folder.

tjwatson committed:
Use proper logs folder for the run action

tjwatson committed:
Merge pull request #18146 from tjwatson/openJ9Update

Open j9 update API and move checkpoint and restore logs to the server logs/ folder

tjwatson committed:
Merge branch 'release' into 'criu' branch

tjwatson committed:
Merge branch 'integration' into criu-merge-integration

tjwatson committed:
Fix test for updates to ExecuteCRIU

tjwatson committed:
Merge pull request #18157 from tjwatson/criu-merge-integration

Criu merge integration

sebratton committed:
criu: "exec restore" checkpoint taken with "server run"

sebratton committed:
Merge pull request #18198 from sebratton/criu-update

criu: "exec restore" checkpoint taken with "server run"

tjwatson committed:
Move checkpoint logs to logs/checkpoint directory

tjwatson committed:
Merge pull request #18199 from tjwatson/criu-checkpoint-logs

Move checkpoint logs to logs/checkpoint directory

Commit:
Only allow one checkpoint to be called

sebratton committed:
checkpoint fat

tjwatson committed:
Add a restore hook for config

This restore hook should detect if any environment
vars have changed that impact configuration values.
The VARIABLE_SOURCE_DIRS value is not considered by
this hook.

tjwatson committed:
Merge pull request #18241 from dettmoney/checkpoint-once

Only allow a single checkpoint to be called

tjwatson committed:
Add a checkpoint action

Checkpoint is now done with the following:

server checkpoint <server-name> --at=features|applications

Restore happens when you issue the run action

server run <server-name>

For now we do not allow for a restore from the start action

tjwatson committed:
Merge pull request #18293 from tjwatson/checkpointAction

Add a checkpoint action

tjwatson committed:
Always write to a console log

This allows for successful restore in the background (server start) even
when the checkpoint is done in the foreground.  When restoring in the
foreground (server run) a tail process is used to display the
console.log to the standard out console.  When restoring the existing
console.log is saved off to checkpoint-console.log and a new console.log
is started.

tjwatson committed:
Merge pull request #18334 from tjwatson/criuConsoleLog

Always write to a console log

tjwatson committed:
Backup workarea for checkpoint, use backup on subsequent restores

When doing a checkpoint the workarea is backed up.  On restore a marker
file is created to indicate that on subsequent restores the workarea
should be restored before doing a criu restore.

This is done to preserve the workarea as it was when the checkpoint was
taken.  An environment variable is introduced (WLP_IN_CONTAINER) that
can be set to true.  When true then the workarea is not backed up on
checkpoint.  The assumption is that the container is always restarted
from the checkpoint image and therefore there is no need to backup the
workarea because the workarea saved in the container image will be
correct.

tjwatson committed:
Make tail use a small sleep amount

In container the tail -f command can be
slow to show the first message.  This gives
the false impression that the server takes longer to
start than in reality.  Shortening the sleep time
improves this perception.

tjwatson committed:
Merge pull request #18358 from tjwatson/backupWorkarea

Backup workarea for checkpoint, use backup on failed restore

tjwatson committed:
Merge branch 'release' into criuMergeRelease

tjwatson committed:
For checkpoint workarea backup, preserve the timestamps

Some caches in liberty are timestamp sensitive.  For example,
the library cache used for spring boot applications seems to
be senstive to any timestamp change of the libraries which
then cause the application to be restarted after a restore
from the backed up workarea.

Doing a cp --preserve=timestamps seems to work around this

tjwatson committed:
Merge pull request #18389 from tjwatson/criuMergeRelease

Merge release and fix timestamps on workarea backup

tjwatson committed:
Properly rotate the checkpoint console log

The checkpoint-console.log was getting overwritten each
restore with the previous console.log.  It really only
should be written to on the first restore.  Subsequent
restores should just remove the existing console log
to start fresh

anjumfatima90 committed:
Use the latest value of VARIABLE_SOURCE_DIRS on retore

sebratton committed:
CRIU: Fix .pid file on restore. Block restore when server running.

sebratton committed:
code review changes

sebratton committed:
Merge pull request #18469 from sebratton/block-restore-on-running-server

CRIU: Fix .pid file on restore. Block restore when server running.

tjwatson committed:
Update to latest Open J9 criu API

sebratton committed:
criu: move workarea backup

move the workarea backup from /workarea-checkpoint to be under the workarea

sebratton committed:
changes from code review for pr 18485

sebratton committed:
Merge pull request #18485 from sebratton/move-workarea-checkpoint-dir

criu: move workarea backup

tjwatson committed:
Merge pull request #18483 from tjwatson/openJ9Updates

Update to latest Open J9 criu API

tjwatson committed:
Add whiteboard for ClassFileTransformer services

This allows a global transformer to be registered that applies
to all AppClassLoaders that get created for the server.

tjwatson committed:
Add deployment phase for checkpoint

The deployment checkpoint happens before executing
the first line of code from an application.
The deployment checkpoint is triggered by code
that is injected into application classes <clinit>
with a ClassFileTransformer.

Once the first <clinit> is done for any application
class then the checkpoint is performed.  When any
checkpoint is performed then the transformer is
unregistered to stop modifying application classes
on the restore side.

sebratton committed:
Criu: check for missing checkpoint-1.0 feature

If a checkpoint operation is requested but the checkpoint feature is missing the checkpoint will be silently ignored. This change is to issue error message stating that the checkpoint feature is missing.

sebratton committed:
Merge pull request #18510 from sebratton/error_missing_checkpoint_feat

criu: issue error message if missing checkpoint-1.0 feature

tjwatson committed:
Add beta guard for the checkpoint action

This is done in LaunchArguments instead of in the server script
because a beta guard check cannot be efficiently done in the server
script.  As a result there is a hardcoded production of the message
about the checkpoint action being unknown.

This is because the server will not use the --help:actions:checkpoint
option when a GA driver is detected.

sebratton committed:
fix feature manager unit test

sebratton committed:
Merge pull request #18546 from sebratton/featureManagerUnitTest

fix feature manager unit test

tjwatson committed:
Delay JMX connector until after ServerStarted service

tjwatson committed:
Merge branch 'release' into `criu` branch

tjwatson committed:
Merge pull request #18632 from tjwatson/mergeRelease

Merge release branch

tjwatson committed:
Split implementation of ExecuteCRIU out for Open J9

Split the ExecuteCRIU out of the activator to
avoid NoClassDefFoundErrors when activating the openj9
bundle.

tjwatson committed:
Merge pull request #18649 from tjwatson/fixOpenJ9Start

Split implementation of ExecuteCRIU out for Open J9

tjwatson committed:
Add beta guard for calls to system transformers

System transformers is part of the checkpoint feature
this should only be enabled for GA once the checkpoint
feature is GA or if another feature GAs before that
needs this functionality.

tjwatson committed:
Merge pull request #18542 from tjwatson/deploymentPhase

Add Deployment phase for checkpoint

sebratton committed:
Criu: narrow scope of trap in bin/server

The trap in bin/server was terminating processes in the same process group as the bash command itself, which can have undesired side effects. Narrow the effect to terminating the tail process (which was the original intent).

sebratton committed:
Merge pull request #18675 from sebratton/bin-server-trap

Criu: narrow scope of trap in bin/server

tjwatson committed:
Fix app start time for deployment phase

If an application has early startup code and the deployment
phase is used then the application start time is recorded at
checkpoint time.  This start time should be restored to
the current nano time.

tjwatson committed:
Merge branch 'release' into 'criu' branch

sebratton committed:
checkpoint: remove jigawatts

Jigawatts criu support not used at this time. Checkpoint will rely on native support for criu in openJ9

tjwatson committed:
Merge pull request #18799 from sebratton/remove-jigawatts

checkpoint: remove jigawatts

tjwatson committed:
Fix TransformerTest to account for system transformers

tjwatson committed:
Merge pull request #18801 from tjwatson/fixUnitTests

Fix TransformerTest to account for system transformers

anjumfatima90 committed:
Handle reprocessConfig differently when called from restore hook and not

tjwatson committed:
Merge pull request #18626 from tjwatson/delayLocalConnector

Delay JMX connector until after ServerStarted service

tjwatson committed:
Merge branch 'release' into 'criu'

tjwatson committed:
Use method references and use common getActualBytes method

tjwatson committed:
Merge pull request #18819 from tjwatson/criuMergeRelease

Merge release into criu branch

tjwatson committed:
Fix unit tests for system transformers

Restructure ClassLoadingServiceImpl to make it easier to unit test beta
functionality with system transformers.

tjwatson committed:
Remove unnecessary @SuppressWarnings("deprecation")

tjwatson committed:
Merge pull request #18812 from tjwatson/restoreAppStartTime

Fix app start time for deployment phase

tjwatson committed:
Add ability to run with a criu stub for debugging

Set the system property io.openliberty.checkpoint.stub.criu=true
to enable a CRIU stub.  This allows for testing of restore hooks
without the need for a real criu support.  It is also helpful
when trying to debug the flow of restore hooks

anjumfatima90 committed:
Merge pull request #18832 from tjwatson/criuStub

Add criu stub

tjwatson committed:
Merge pull request #18826 from tjwatson/fixClassLoadingUnitTest

Fix unit tests for system transformers

tjwatson committed:
Pass the class name instead of the resource name to transformers

The refactoring of the transformer code broke by passing the
resource name to the transformers.

tjwatson committed:
Merge pull request #18884 from tjwatson/criuFixTransformers

Pass the class name instead of the resource name to transformers

tjwatson committed:
Merge pull request #18970 from anjumfatima90/env_var_change

Add restore hook to reprocess config when environment variable changes

tjwatson committed:
Merge branch 'release' into criu

tjwatson committed:
Merge pull request #19038 from tjwatson/criuMergeRelease

Criu merge release

anjumfatima90 committed:
Save the variable source dirs in cache

sebratton committed:
Merge pull request #19039 from sebratton/criu-fat-min

Criu fat tests

tjwatson committed:
Do not capture standard in when doing a checkpoint

There is no need to capture standard in when doing a
checkpoint.  This also helps for restore if we can
avoiding having to connect standard in to a TTY
for some restore scenarios

sebratton committed:
Test for criu via a seperate jvm

sebratton committed:
criu support check tied to machine

tjwatson committed:
Merge pull request #19126 from tjwatson/checkpointStdIn

Do not capture standard in when doing a checkpoint

sebratton committed:
Merge pull request #19147 from sebratton/criu-support-probe

Check  vm/platform support for criu in fat via fork of new jvm.

sebratton committed:
Handle CRIU not supported scenarios

Consolidate checks for criu support to a single class and add a check for criu not installed on platform

tjwatson committed:
Merge pull request #19188 from sebratton/criu-notsupported-scenarios

Handle CRIU not supported scenarios

tjwatson committed:
Update the CRIUSupport API

New method added to restore the env

tjwatson committed:
Move openj9 based impl of ExecuteCRIU into checkpoint

Moving ExecuteCRIU_OpenJ9 into io.openliberty.checkpoint
to simplify the code.  No longer need to allow alternative
implementations and it is easier to avoid a service lookup
for ExecuteCRIU implementations.

tjwatson committed:
Set criu stub at construction time

Add default methods for ExecuteCRIU to do nothing

tjwatson committed:
Merge pull request #19195 from tjwatson/rmCheckpointJ9Bundle

Move openj9 based impl of ExecuteCRIU into checkpoint

tjwatson committed:
Have CRIUSupport set the restored environment variables

A new file .env.properties is used to store the restored environment in.
CRIUSupport then reads this file at restore time to set any new
environment values.  Care must be taken to not set different values for
the variables set at checkpoint time.  Otherwise CRIUSupport will fail
to restore with an exception.

sebratton committed:
criu FAT changes

Add lambda callback and independent checkpoint/restore functions. Refactor.

sebratton committed:
Code review suggestion

sebratton committed:
Merge branch 'release' into criu-fat

tjwatson committed:
When restoring calculate variable changes do to env value changes

sebratton committed:
fix libertyServer startserver logic for non checkpoint operation

tjwatson committed:
Merge branch 'release' into 'criu'

sebratton committed:
Merge pull request #18807 from sebratton/criu-fat

Merged release branch to criu and added checkpoint fat changes

sebratton committed:
criu: additional fats. Changes to LibertyServer

Added a couple of fat tests. Modified LibertyServer to allow negative test where checkpoint failed

sebratton committed:
Merge pull request #19333 from sebratton/criu-fat

Add criu fat

tjwatson committed:
Register the checkpoint Phase as a service.

sebratton committed:
criu Add skip annotation

add annotation to the fat framework to allow skipping a test when criu is not supported

sebratton committed:
Merge pull request #19350 from sebratton/criu-fat

criu Add skip annotation

tjwatson committed:
Merge pull request #19349 from tjwatson/testRebase

Ability to restore the environment variables

tjwatson committed:
Enable CheckpointImpl based on the Phase service

tjwatson committed:
Add LOG related env to the list of excluded variables to restore

tjwatson committed:
Merge branch integration into criu

tjwatson committed:
Call restore hooks in reverse order of prepare hooks

tjwatson committed:
Set Phase service property to the current phase.

tjwatson committed:
Add hook to prepare and restore the OSGi console

The console does not handle keeping its telnet port active through a
checkpoint/restore operation.  This hook stops/refreshes the equinox
console bundle on checkpoint and then restarts it on restore to
re-establish the OSGi console port.

tjwatson committed:
Ensure the checkpoint fat has at least one test run

tjwatson committed:
Merge pull request #19390 from tjwatson/phaseService

Register the Phase as a service

tjwatson committed:
Remove CheckpointHookFactory

If a hook needs to know what phase is being checkpoint/restore they can
look for the CheckpointPhase service enum.

anjumfatima90 committed:
Fix the failure caused during restore when messages.log is deleted

Add a checkpoint hook in FrameworkManager to stop the log provider during prepare and then update it during restore

anjumfatima90 committed:
Set service ranking to min value and update TrConfiguartor with a singleton map on restore

anjumfatima90 committed:
Register logger handler and capture system streams on restore

anjumfatima90 committed:
Fix method arguments

anjumfatima90 committed:
Update copyright header year

tjwatson committed:
Merge pull request #19339 from anjumfatima90/issue18395-CheckpointHookToStopAndUpdateLogProvider

Fix the failure caused during restore when messages.log is deleted

tjwatson committed:
Run the TimestampUtils restore hook run early

restore hooks are run in opposite order of prepare hooks.  Setting
the service ranking to be very low ensures this restore hook
will be run early.

tjwatson committed:
Print the timed exit message on restore

The FAT server launcher expects to see a message
aboout timed exit being loaded.  This restore hook
simply makes sure the message is posted on restoring
a server from checkpoint

tjwatson committed:
Ensure messages about application starting happen on restore

When restoring a checkpoint server that has already started
applications the application starting messages will not
be posted for each server restore run.

This could be confusing to users not to see messages
indicating their application is started.  It also poses
issues for our FAT tests that verify the configured
applications are started.

To address this the ApplicationConfigurator checkpoint
hook is enhanced to collect application messages that
happen pre-checkpoint and post them on resore.

tjwatson committed:
Merge branch 'release' into 'criu'

tjwatson committed:
Merge pull request #19732 from tjwatson/mergeBranch

Merge release into criu branch

anjumfatima90 committed:
Add test to test variable source dirs update during restore

anjumfatima90 committed:
Add service ranking

jantley-ibm commented on dev/com.ibm.ws.kernel.boot/resources/com/ibm/ws/kernel/boot/resources/LauncherMessages.nlsprops:
```diff
@@ -708,6 +708,10 @@ warning.earlyRelease=CWWKE0953W: This version of {0} is an unsupported early rel
 warning.earlyRelease.useraction=Do not use an early release version on production systems, since early release versions are not supported.
 
+error.invalidPhaseName=CWWKE0954E: An empty or unknown checkpoint phase "{0}" was specified.
+error.invalidPhaseName.explanation=An unknown phase was passed in as the value of the checkpoint option.
```
change
passed in
to
specified

tjwatson replied:
Should be fixed now.

jantley-ibm commented on dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops:
```diff
@@ -0,0 +1,30 @@
+# MSG_DESCRIPTIVE_NAME_CWWKC0000.explanation=Explanation text for the message
+# MSG_DESCRIPTIVE_NAME_CWWKC0000.useraction=User action text for the message
+
+CHECKPOINT_DUMP_INITIATED_CWWKC0451=CWWKC0451I: A server checkpoint was requested. The server will stopped when the checkpoint completes.
```
change
The server will stopped
to
The server will stop

tjwatson replied:
Fixed.

jantley-ibm commented on dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops:
```diff
@@ -0,0 +1,30 @@
+
+CHECKPOINT_DUMP_INITIATED_CWWKC0451=CWWKC0451I: A server checkpoint was requested. The server will stopped when the checkpoint completes.
+CHECKPOINT_DUMP_INITIATED_CWWKC0451.explanation=A checkpoint was requested. The server is halted and its current state is written to disk.
+CHECKPOINT_DUMP_INITIATED_CWWKC0451.useraction=Check the knowledge center for details of the checkpoint/restore operations.
```
change
knowledge center
to
documentation

tjwatson replied:
Fixed.

jantley-ibm commented on dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops:
```diff
@@ -0,0 +1,30 @@
+CHECKPOINT_DUMP_INITIATED_CWWKC0451.explanation=A checkpoint was requested. The server is halted and its current state is written to disk.
+CHECKPOINT_DUMP_INITIATED_CWWKC0451.useraction=Check the knowledge center for details of the checkpoint/restore operations.
+#
+CHECKPOINT_RESTORE_CWWKC0452I=CWWKC0452I: The liberty server is resuming execution from a checkpoint.
```
change
liberty server is resuming execution
to
Liberty server is resuming operation

tjwatson replied:
Fixed.

jantley-ibm commented on dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops:
```diff
@@ -0,0 +1,30 @@
+CHECKPOINT_DUMP_INITIATED_CWWKC0451.useraction=Check the knowledge center for details of the checkpoint/restore operations.
+#
+CHECKPOINT_RESTORE_CWWKC0452I=CWWKC0452I: The liberty server is resuming execution from a checkpoint.
+CHECKPOINT_RESTORE_CWWKC0452I.explanation=A restore operation was run on a liberty server checkpoint image.
```
change
liberty
to
Liberty

tjwatson replied:
Fixed.

jantley-ibm commented on dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops:
```diff
@@ -0,0 +1,30 @@
+#
+CHECKPOINT_RESTORE_CWWKC0452I=CWWKC0452I: The liberty server is resuming execution from a checkpoint.
+CHECKPOINT_RESTORE_CWWKC0452I.explanation=A restore operation was run on a liberty server checkpoint image.
+CHECKPOINT_RESTORE_CWWKC0452I.useraction=Check the knowledge center for details of the checkpoint/restore operations.
```
change
knowledge center
to
documentation

tjwatson replied:
Fixed.

jantley-ibm commented on dev/com.ibm.ws.kernel.feature.core/resources/com/ibm/ws/kernel/feature/internal/resources/ProvisionerMessages.nlsprops:
```diff
@@ -231,6 +231,11 @@ UPDATE_INDIRECT_CONFLICT_INCOMPATIBLE_FEATURES_SAME_PLATFORM_ERROR=CWWKF0047E: T
 UPDATE_INDIRECT_CONFLICT_INCOMPATIBLE_FEATURES_SAME_PLATFORM_ERROR.useraction=Try to specify a compatible version of the configured features that caused the conflict in the server.xml file. If no other versions are available, then the configured features are not compatible with each other and cannot be configured at the same time.
 
+#CWWKF0048E
+CHECKPOINT_REQUESTED_CHECKPOINT_FEATURE_MISSING=CWWKF0048E: The server command specified the checkpoint action. A checkpoint cannot be taken because the checkpoint-1.0 feature was not configured in the server.xml file.
```
change
was not
to
is not

tjwatson replied:
Fixed.

jantley-ibm commented on dev/com.ibm.ws.kernel.feature.core/resources/com/ibm/ws/kernel/feature/internal/resources/ProvisionerMessages.nlsprops:
```diff
@@ -231,6 +231,11 @@ UPDATE_INDIRECT_CONFLICT_INCOMPATIBLE_FEATURES_SAME_PLATFORM_ERROR=CWWKF0047E: T
 
+#CWWKF0048E
+CHECKPOINT_REQUESTED_CHECKPOINT_FEATURE_MISSING=CWWKF0048E: The server command specified the checkpoint action. A checkpoint cannot be taken because the checkpoint-1.0 feature was not configured in the server.xml file.
+CHECKPOINT_REQUESTED_CHECKPOINT_FEATURE_MISSING.explanation=The server checkpoint action requires that the checkpoint-1.0 feature be configured.  
```
Change to
The checkpoint-1.0 feature must be configured to use a server checkpoint action.

tjwatson replied:
Fixed

jantley-ibm reviewed (commented):
ID edits provided, let me know if you have questions. I will add the "ID reviewed" tag when complete.

tjwatson committed:
Merge branch release into criu

tjwatson committed:
Adjust unit test for FrameworkManager changes

tjwatson committed:
Merge pull request #19735 from tjwatson/removeCheckpointHookFactory

Remove checkpoint hook factory

tjwatson committed:
Update OpenJ9 CRIU API with support for prepare and restore hooks

Now we can pass prepare and restore Runnables to the CRIUSupport so that
our prepare and restore hooks will be run by the JVM in single threaded
mode.

The JVM also changed how isCRIUSupportedEnabled works by having it
always return true or false and never throwing an exception.  Instead we
now have a method CRIUSupport.getErrorMessage() that can tell us why
CRIU is not supported.

tjwatson committed:
Revert "Initial drop of getAvailableProcessors changes"

This reverts commit da81e6b14e16c088fe0e5c38f0252ca77b68bd44.

This revert only restores the ConcurrentHashSet impl in util.
All other changes are left.

anjumfatima90 committed:
Refactor registering and unregistering of service registrations

tjwatson committed:
Merge branch 'integration' into 'criu' branch

tjwatson committed:
Adjust impl to criu branch changes

tjwatson committed:
Merge pull request #19773 from tjwatson/mergeBranch

Merge integration into criu

tjwatson committed:
Fix up messages from ID review

tjwatson committed:
Ensure the exit code is not zero when checkpoint fails.

tjwatson committed:
Log an error message when checkpoint or restore fail

tjwatson committed:
Merge pull request #19759 from tjwatson/passJ9PrepareRestoreRunnable

Update OpenJ9 CRIU API with support for prepare and restore hooks

tjwatson committed:
Merge pull request #19821 from tjwatson/idReview

Fix up messages from ID review

tjwatson commented:
#build

tjwatson commented:
@dmuelle thanks for doing the review of the initial messages for checkpoint.  I think I have addressed your comments.  In addition we have added more messages to `dev/io.openliberty.checkpoint/resources/io/openliberty/checkpoint/resources/CheckpointMessages.nlsprops`.  Please have a fresh look, thanks.

sebratton committed:
checkpoint: fix testCheckpointFeatureMissingError in FAT

sebratton committed:
Merge pull request #19839 from sebratton/criu

checkpoint: fix testCheckpointFeatureMissingError in FAT

tjwatson committed:
Some hooks may required a JVM in multi-threaded mode

By default the hooks prepare and restore hooks are run
while the JVM is in single-threaded mode.  Some hooks
may require that they be running while the JVM is in
multi threaded mode.  This change adds a service
property for a hook to specify that they must be
run while the JVM is in multi-threaded mode.

The CheckpointOSGiConsole hook has been updated to be run
while the JVM is in multi-threaded mode and a new FAT
test has been added to ensure the osgiConsole-1.0 feature
is functional for checkpoint/restore

tjwatson committed:
Remove unused BundleContext parameter

tjwatson committed:
Merge pull request #18413 from anjumfatima90/variable_source_dir

Restore hook for config change

tjwatson committed:
Remove unused jigawatts coordinates

tjwatson committed:
Remove unused criu-ns script

tjwatson committed:
Merge pull request #19848 from tjwatson/multiThreadedHooks

Allow hooks to run while the JVM is in multi-thread mode

skasund commented:
L2 message review not required.

tjwatson commented:
Closing this in favor of the rebase/squash approach in PR #19856
